### DISCUSSION TITLE
Change logName before creating

### DISCUSSION THREAD

**Discussion by argo-uka** [spatie/laravel-activitylog#958] [Q&A] [answered]
Hello,

I take advantage of the LogsActivity trait to automatically log when creating models. But I have different places where I create new Model and I would like to set a different logName depending on that.
Example:
```php
public function fromMethodA () {
    activity("create-from-a");
   MyModel::create(.....);
}

public function fromMethodB () {
    activity("create-from-b");
   MyModel::create(.....);
}
```

So far I tried to set `$activity->log_name` with `tap`, but it doesn't work. Any idea on how to achieve this?

**Comment by Gummibeer** [comment, marked as answer]
Hey, you could define a `tapActivity()` method on your model which can change the activity before it's persisted. But you need some fancy way to transfer the calling method to this tap method.
Without any container usage or similar global variable holder you could use `debug_backtrace()` for example.

Another option would be

```php
$default = config('activitylog.default_log_name');
config('activitylog.default_log_name', 'my_new_log_name');
$model = MyModel::create(...);
config('activitylog.default_log_name', default');

return $model;
```

You could make this more readable and reusable by adding a macro to the `ActivityLogger`:

```php
ActivityLogger::macro('withLogName', function(string $logName, Closure $callback) {
    $defaultLogName = config('activitylog.default_log_name');
    config('activitylog.default_log_name', $logName');
    $result = $callback();
    config('activitylog.default_log_name', default');

    return $result;
});
```

And you could use it like the following:

```php
$model = ActivityLogger::withLogName('my_new_log_name', function() {
    return MyModel::create(...);
});
```

**Reply by argo-uka** [reply]
Amazing, thank you for the quick reply. I had to change a little bit. 

```php
ActivityLogger::macro('withLogName', function(string $logName, Closure $callback) {
  $defaultLogName = config('activitylog.default_log_name');
  config(['activitylog.default_log_name' => $logName]);
  $result = $callback();
  config(['activitylog.default_log_name' => $defaultLogName]);
  return $result;
 });
```
